doc/source/reference/c-api/array.rst

```diff
@@ -1555,8 +1555,296 @@ For all of these macros *arr* must be an instance of a (subclass of)
+
+       type_in in_value;
+       type_out out_value
+       memcpy(&value, in_data, sizeof(type_in));
+       out_value = in_value;
+       memcpy(out_data, &out_value, sizeof(type_out)
+
+   while a normal loop can just use::
+
+       *(type_out *)out_data = *(type_in)in_data;
+
+   The unaligned loops are currently only used in casts and will never be picked
+   in ufuncs (ufuncs create a temporary copy to ensure aligned inputs).
+   These flags are ignored when ``NPY_METH_get_loop`` is defined, where instead
+   whichever loop returned by the ``get_loop`` function is used.
+
+.. c:macro:: NPY_METH_contiguous_indexed_loop
+
+   A specialized inner-loop option to speed up common ``ufunc.at`` computations.
+
+.. c:macro:: NPY_METH_get_loop
+
+   Allows more fine-grained control over loop selection. Accepts an
+   implementation of PyArrayMethod_GetLoop, which in turn returns a
+   strided loop implementation. If ``NPY_METH_get_loop`` is defined,
+   the other loop slot IDs are ignored, if specified.
+
+Flags
+~~~~~
+
+.. c:enum:: NPY_ARRAYMETHOD_FLAGS
+
+   These flags allow switching on and off custom runtime behavior for
+   ArrayMethod loops.  For example, if a ufunc cannot possibly trigger floating
+   point errors, then the ``NPY_METH_NO_FLOATINGPOINT_ERRORS`` flag should be
+   set on the ufunc when it is registered.
+
+   .. c:enumerator:: NPY_METH_REQUIRES_PYAPI
+
+      Indicates the method must hold the GIL. If this flag is not set, the GIL
+      is released before the ufunc is called.
+
+   .. c:enumerator:: NPY_METH_NO_FLOATINGPOINT_ERRORS
+
+      Indicates the method cannot generate floating errors, so checking for
+      floating errors after the ufunc completes can be skipped.
+
+   .. c:enumerator:: NPY_METH_SUPPORTS_UNALIGNED
+
+      Indicates the method supports unaligned access.
+
+   .. c:enumerator:: NPY_METH_IS_REORDERABLE
+
+      Indicates that the result of applying the loop repeatedly (for example, in
+      a reduction operation) does not depend on the order of application.
+
+   .. c:enumerator:: NPY_METH_RUNTIME_FLAGS
+
+      The flags that can be changed at runtime.
+
+.. _arraymethod-typedefs:
+
+Typedefs
+~~~~~~~~
+
+Typedefs for functions that users of the ArrayMethod API can implement are
+described below.
+
+.. c:type:: NPY_CASTING (PyArrayMethod_ResolveDescriptors)( \
+                struct PyArrayMethodObject_tag *method, \
+                PyArray_DTypeMeta **dtypes, \
+                PyArray_Descr **given_descrs, \
+                PyArray_Descr **loop_descrs, \
+                npy_intp *view_offset)
+
+   The resolve descriptors function. Given the DType classes the method was
+   created for and the concrete descriptor instances used at runtime, determines
+   the correct descriptors to pass into the loop implementation, which may or
+   may not be the descriptors passed in by the user. The *method* is an opaque
+   pointer to the underlying cast or ufunc loop which is currently exposed
```

Comment:
what does "which is currently exposed publicly" mean here? I would have expected a (link to a) description of how to use this inside one's own resolution loop.